lib/internal/fs/dir.js

```diff
@@ -48,6 +54,42 @@ class Dir {
     return this[kDirPath];
   }
 
+  [kDirGetBufferedDirEnt](encoding) {
+    // We end up in this function if we have previously already read entries
+    // from disk.
+    // There is some complexity added here through the fact that .read()
+    // and .readSync() may theoretically be called with different `encoding`
+    // values at different invocations.
+    // The `kDirBufferedEntries` field of this object contains a list of tuples
+    // describing directory entries that were already read during a previous
+    // `.read()` call. Each tuple has either length 2 or 3; the format being
+    // [ name : string, type : number ] or
+    // [ name : string , nameAsBuffer : Buffer, type : number ].
+    // The 2-valued format is used if the original Buffer for the name can
+    // be reconstructed from the name string, should another encoding than the
+    // one for the original read be requested.
+    const bufferedEntries = this[kDirBufferedEntries];
+    const hasNameAsBuffer = typeof bufferedEntries[1] !== 'number';
+    const result = bufferedEntries.splice(0, hasNameAsBuffer ? 3 : 2);
```

Comment:
Would it be possible to reverse the tuples? That way the entries could be popped from the array and that could be another performance improvement.

Reply:
I guess so… although with the maximum # of entries being 32, I wouldn’t think there’s a huge performance cliff here? Like, if this does move things in memory, it is going to to so in a very limited way.

Reply:
In theory it's up to 32 * 3 but that's of course still a low number. The memory copy and toString operations are likely the most costly operations here. It was just a potential minor improvement that came to my mind. Nothing that has to be followed up upon.

Thinking about the costly operations: is it possible to create some benchmark test data for the worst case scenario as well?